packages/frint/src/App.ts

```diff
@@ -0,0 +1,410 @@
+    if (!regionName) {
+      return this._apps$;
+    }
+
+    return this._apps$
+      .pipe(map$((collection) => {
+        return collection
+          .filter((w) => {
+            return w.regions.indexOf(regionName) > -1;
+          });
+      }));
+  }
+
+  public registerApp(appClass: AppClass, opts: RegisterAppOptions = {}) {
+    const options = {
+      multi: false,
+      ...opts,
+    };
+
+    if (typeof options.name !== 'undefined') {
+      Object.defineProperty(appClass, 'frintAppName', {
+        value: options.name,
+        configurable: true,
+      });
+    }
+
+    const existingIndex = findIndex(this._appsCollection, (w) => {
+      return w.name === appClass.frintAppName;
+    });
+
+    if (existingIndex !== -1) {
+      throw new Error(`App '${appClass.frintAppName}' has been already registered before.`);
+    }
+
+    this._appsCollection.push({
+      ...options,
+      name: appClass.frintAppName,
+      appClass,
+      instances: {},
+      regions: options.regions || [],
+    });
+
+    if (options.multi === false) {
+      this.instantiateApp(appClass.frintAppName);
+    }
+
+    this._apps$.next(this._appsCollection);
+  }
+
+  public hasAppInstance(name, region = null, regionKey = null) {
+    const instance = this.getAppInstance(name, region, regionKey);
+
+    if (instance && typeof instance !== 'undefined') {
+      return true;
+    }
+
+    return false;
+  }
+
+  public getAppInstance(name, region = null, regionKey = null) {
+    const index = findIndex(this._appsCollection, a => {
+      return a.name === name;
+    });
+
+    if (index === -1) {
+      return null;
+    }
+
+    const app = this._appsCollection[index];
+    const instanceKey = makeInstanceKey(region, regionKey, app.multi);
+    const instance = app.instances[instanceKey];
+
+    if (!instance || typeof instance === 'undefined') {
+      return null;
+    }
+
+    return instance;
+  }
+
+  public getAppOnceAvailable$(name, region = null, regionKey = null) {
+    const rootApp = this.getRootApp();
+
+    const w = rootApp.getAppInstance(name, region, regionKey);
+
+    if (w) {
+      return Observable.of(w);
+    }
+
+    return rootApp._apps$
+      .pipe(
+        concatMap$(y => y),
+        find$(app => app.name === name),
+        map$((x) => {
+          const instanceKey = makeInstanceKey(region, regionKey, x.multi);
+          return x.instances[instanceKey];
+        }),
+        first$()
+      );
+  }
+
+  public instantiateApp(name, region = null, regionKey = null) {
+    const index = findIndex(this._appsCollection, a => {
+      return a.appClass.frintAppName === name;
+    });
+
+    if (index === -1) {
+      throw new Error(`No app found with name '${name}'.`);
+    }
+
+    const w = this._appsCollection[index];
+    const key = makeInstanceKey(region, regionKey, w.multi);
+
+    this._appsCollection[index].instances[key] = new w.appClass({
+      ...omit(w, ['AppClass', 'instances']) as AppOptions,
```

Comment:
This is also not too nice. We're dynamically cherrypicking some of the properties of the object, so `tsc` can not figure out the type of the result, so I have to cast to `AppOptions` explicitly.

Reply:
you are right, this part is not very well thought out.

we can combine efforts when we tackle #182. the API for internal registry of Child Apps can be done the right way then.